Neighborhood attention transformer (#88)
creatis-myriad/creatis-myriad.github.io: collections/_posts/2022-11-02-NeighborhoodAttentionTransformer.md

---
layout: review
title: "Neighborhood Attention Transformer"
tags: deep-learning CNN transformer segmentation classification object-detection attention
author: "Pierre Rougé"
cite:
    authors: "Ali Hassani, Steven Walton, Jiachen Li, Shen Li, Humphrey Shi"
    title:   "Neighborhood Attention Transformer"

> pdf: "https://arxiv.org/abs/2204.07143v1"

Reviewer: I typicall prefer not to refer to a specific version of an arxiv paper (the `v1` at the end here) so that URL always leads to the most recent version.

> ---

Reviewer: The venue where the paper was published is missing. If the paper has only been published on arXiv as of yet, put arXiv as venue. That way, it's clear the paper is available as a pre-print and it's not just an oversight.

# Notes

* Code is available on [github](https://github.com/SHI-Labs/Neighborhood-Attention-Transformer)
* This work was done by the same team that did the Compact Convolutional Transformer (CCT) reviewed in this [post](https://creatis-myriad.github.io./2022/06/13/CompactConvolutionalTransformer.html)
* So they use the same method of Convolutional Tokenization

# Highlights

* Similar to Swin Transformer the idea is to reduce the computational cost of the attention mechanism 
* The authors introduce the Neighborhood Attention (NA) and the Neighborhood Attention Transformer (NAT)
* **With the Neighborhood Attention the attention is only computed on a neighborhood around each token**
* This method not only allows to reduce the computational cost of the attention mechanism but also helps to introduce local inductive biases

> *  The drawback is that it reduces the receptive field

Reviewer: Double space here:
```suggestion
* The drawback is that it reduces the receptive field
```

![](/collections/images/NeighborhoodAttentionTransformer/receptive_fields.jpg)

# Neighborhood Attention

![](/collections/images/NeighborhoodAttentionTransformer/NeighborhoodAttention.jpg)

> * Neighborhood attention on a single pixel $$(i, j)$$ is defined as follows:

Reviewer: I don't think the list here is necessary/helpfull:
```suggestion
Neighborhood attention on a single pixel $$(i, j)$$ is defined as follows:
```

$$ NA(X_{i, j}) = softmax(\frac{Q_{i,j}K^T_{\rho(i,j)} + B_{i,j}}{scale})V_{\rho_{(i,j)}} $$

where $$Q, K, V$$ are linear projections of $$X$$

$$B_{i,j}$$ denotes the relative positional bias

with ρ(i, j), which is a fixed-length set of indices of pixels nearest to (i, j)

for a neighborhood of size $$ L * L $$, $$ \lVert \rho(i,j) \rVert = \lVert L² \rVert$$	

> > However, if the function ρ maps each pixel to all pixels ($$L²$$is equal to feature map size), this will be equivalent to self attention.

Reviewer: Missing space:
```suggestion
> However, if the function ρ maps each pixel to all pixels ($$L²$$ is equal to feature map size), this will be equivalent to self attention.
```

- The complexity of the neighborhood attention is linear with respect to resolution unlike self attention's.
- The function $$\rho$$ which maps a pixel to a set of neighboring pixels is realized with a sliding window.
- For corner pixels that cannot be centered, the neighborhood is expanded to maintain receptive field size. As illustrated in the image below.

![](/collections/images/NeighborhoodAttentionTransformer/corner_pixels.jpg)

# Neighborhood Attention Transformer

![](/collections/images/NeighborhoodAttentionTransformer/architecture.jpg)

- For the tokenization they use the overlapping convolution method introduced in the  [Compact Convolutional Transformer](https://creatis-myriad.github.io./2022/06/13/CompactConvolutionalTransformer.html)

- The rest of the architecture is a succession of blocks containing a token merging layer to reduce dimension and a standard multi head attention block but with the self attention replaced by the neighborhood attention

- The token merging layer is also different from the patch merging layer in the Swin Transformer

> - Here the overlapping downsampler consists in a convolution 3 x 3 with strides 2 x 2 on the patches

Reviewer: I think it's better with no space between the kernel sizes :wink: 
```suggestion
- Here the overlapping downsampler consists in a convolution 3x3 with strides 2x2 on the patches
```

# Results

## Classification

- Trained on ImageNet-1k (1.2 millions images for training, 1000 classes)

![](/collections/images/NeighborhoodAttentionTransformer/results_classification.jpg)

> - NAT outperforms significantly Swin Transformers and ConvNeXt

Author: I agree, I changed this part. Thank you for the review Nathan =)

* So they use the same method of Convolutional Tokenization

# Highlights

* Similar to Swin Transformer the idea is to reduce the computational cost of the attention mechanism 
* The authors introduce the Neighborhood Attention (NA) and the Neighborhood Attention Transformer (NAT)
* **With the Neighborhood Attention the attention is only computed on a neighborhood around each token**
* This method not only allows to reduce the computational cost of the attention mechanism but also helps to introduce local inductive biases
* The drawback is that it reduces the receptive field

![](/collections/images/NeighborhoodAttentionTransformer/receptive_fields.jpg)

# Neighborhood Attention

![](/collections/images/NeighborhoodAttentionTransformer/NeighborhoodAttention.jpg)

Neighborhood attention on a single pixel $$(i, j)$$ is defined as follows:

$$ NA(X_{i, j}) = softmax(\frac{Q_{i,j}K^T_{\rho(i,j)} + B_{i,j}}{scale})V_{\rho_{(i,j)}} $$

where:

-  $$Q, K, V$$ are linear projections of $$X$$

- $$B_{i,j}$$ denotes the relative positional bias

> - ρ(i, j) is a fixed-length set of indices of pixels nearest to (i, j), for a neighborhood of size $$ L * L $$, $$ \lVert \rho(i,j) \rVert = \lVert L² \rVert$$	

Reviewer: Split the sentence + the operator for the cardinality of a set is just a single vertical line, not the double one:
```suggestion
- ρ(i, j) is a fixed-length set of indices of pixels nearest to (i, j). For a neighborhood of size $$ L * L $$, $$ \vert \rho(i,j) \vert = L^2$$
```

- The complexity of the neighborhood attention is linear with respect to resolution unlike self attention's.
- The function $$\rho$$ which maps a pixel to a set of neighboring pixels is realized with a sliding window.
- For corner pixels that cannot be centered, the neighborhood is expanded to maintain receptive field size. As illustrated in the image below.

![](/collections/images/NeighborhoodAttentionTransformer/corner_pixels.jpg)

# Neighborhood Attention Transformer

![](/collections/images/NeighborhoodAttentionTransformer/architecture.jpg)

- For the tokenization they use the overlapping convolution method introduced in the  [Compact Convolutional Transformer](https://creatis-myriad.github.io./2022/06/13/CompactConvolutionalTransformer.html)

- The rest of the architecture is a succession of blocks containing a token merging layer to reduce dimension and a standard multi head attention block but with the self attention replaced by the neighborhood attention

- The token merging layer is also different from the patch merging layer in the Swin Transformer

> - Here the overlapping downsampler consists in a convolution 3x3 with strides 2x2 on the patches

Reviewer 2: ```suggestion
- Here the overlapping downsampler consists of a 3x3 convolution with 2x2 strides on the patches
```

# Results

## Classification

- Trained on ImageNet-1k (1.2 millions images for training, 1000 classes)

![](/collections/images/NeighborhoodAttentionTransformer/results_classification.jpg)

- NAT outperforms Swin Transformers and ConvNeXt

## Object Detection

- Mask R-CNN and Cascade Mask R-CNN with different backbones trained on MS-COCO

  ![](/collections/images/NeighborhoodAttentionTransformer/results_object_detection.jpg)

## Semantic Segmentation

- UPerNet with different backbones trained on ADE20K (20 000 training images)

![](/collections/images/NeighborhoodAttentionTransformer/results_semantic_segmentation.jpg)

- NAT performs better than Swin Transformer for the segmentation task
- But NAT fails to beat ConvNeXt, a recent and very efficient convolutional network 

## Ablation studies

> - To attests the efficiency of the Neighborhood Attention they test their architecture on ImageNet-1k with different kinds of attention

Reviewer 2: ```suggestion
- To attest the efficiency of the Neighborhood Attention, they test their architecture on ImageNet-1k with different kinds of attention
```
